libsawtooth/src/journal/genesis.rs

```diff
@@ -0,0 +1,638 @@
+                    error!("Unable to send scheduler error; receiver must have dropped");
+                }
+            }))
+            .map_err(|err| {
+                GenesisError::BatchValidationError(format!(
+                    "Unable to set error callback on scheduler: {}",
+                    err
+                ))
+            })?;
+
+        debug!("Adding {} genesis batches", genesis_data.batches().len());
+
+        for batch in genesis_data.batches() {
+            scheduler.add_batch(batch.clone()).map_err(|err| {
+                GenesisError::BatchValidationError(format!(
+                    "While adding a batch to the schedule: {:?}",
+                    err
+                ))
+            })?;
+        }
+
+        scheduler.finalize().map_err(|err| {
+            GenesisError::BatchValidationError(format!(
+                "During call to scheduler.finalize: {:?}",
+                err
+            ))
+        })?;
+
+        self.transaction_executor
+            .submit(
+                scheduler.take_task_iterator().map_err(|_| {
+                    GenesisError::BatchValidationError(
+                        "Unable to take task iterator from scheduler".to_string(),
+                    )
+                })?,
+                scheduler.new_notifier().map_err(|_| {
+                    GenesisError::BatchValidationError(
+                        "Unable to get new notifier from scheduler".to_string(),
+                    )
+                })?,
+            )
+            .map_err(|err| {
+                GenesisError::BatchValidationError(format!(
+                    "During call to ExecutionTaskSubmitter.submit: {}",
+                    err
+                ))
+            })?;
+
+        // collection results from the scheduler
+        let mut execution_results = vec![];
+        loop {
+            match result_rx.recv() {
+                Ok(SchedulerEvent::Result(result)) => execution_results.push(result),
+                Ok(SchedulerEvent::Complete) => break,
+                Ok(SchedulerEvent::Error(err)) => {
+                    return Err(GenesisError::BatchValidationError(format!(
+                        "During execution: {:?}",
+                        err
+                    )))
+                }
+                Err(err) => {
+                    return Err(GenesisError::BatchValidationError(format!(
+                        "Error while trying to receive scheduler event: {:?}",
+                        err
+                    )))
+                }
+            }
+        }
+
+        // Collect reciepts and changes from execution results
+        let mut receipts = vec![];
+        let mut changes = vec![];
+        for batch_result in execution_results {
+            if !batch_result.receipts.is_empty() {
+                for receipt in batch_result.receipts {
+                    match receipt.transaction_result.clone() {
```

Comment:
Is this clone necessary? Seems like it shouldn't be, but if it is, maybe it can be pushed down so we're only cloning the specific data that's needed (rather than the invalid result data, for instance, which isn't used at all).